test/unit/app/controllers/permissions/permissions-controller-test.js

```diff
@@ -0,0 +1,1387 @@
+
+      permController.removePermissionsFor({
+        [ORIGINS.c]: [PERM_NAMES.eth_accounts],
+      })
+
+      const aAccounts = await permController.getAccounts(ORIGINS.a)
+      const bAccounts = await permController.getAccounts(ORIGINS.b)
+
+      assert.deepEqual(
+        aAccounts, ACCOUNT_ARRAYS.a,
+        'first origin does not have correct accounts'
+      )
+      assert.deepEqual(
+        bAccounts, ACCOUNT_ARRAYS.b,
+        'second origin does not have correct accounts'
+      )
+
+      assert.deepEqual(
+        Object.keys(permController.permissions.getDomains()),
+        [ORIGINS.a, ORIGINS.b],
+        'should have correct domains'
+      )
+    })
+  })
+
+  describe('validatePermittedAccounts', function () {
+
+    let permController
+
+    beforeEach(function () {
+      permController = initPermController()
+      grantPermissions(
+        permController, ORIGINS.a,
+        PERMS.finalizedRequests.eth_accounts(ACCOUNT_ARRAYS.a)
+      )
+      grantPermissions(
+        permController, ORIGINS.b,
+        PERMS.finalizedRequests.eth_accounts(ACCOUNT_ARRAYS.b)
+      )
+    })
+
+    it('throws error on non-array accounts', async function () {
+
+      assert.rejects(
+        permController.validatePermittedAccounts(undefined),
+        ERRORS.validatePermittedAccounts.invalidParam(),
+        'should throw on undefined'
+      )
+
+      assert.rejects(
+        permController.validatePermittedAccounts(false),
+        ERRORS.validatePermittedAccounts.invalidParam(),
+        'should throw on false'
+      )
+
+      assert.rejects(
+        permController.validatePermittedAccounts(true),
+        ERRORS.validatePermittedAccounts.invalidParam(),
+        'should throw on true'
+      )
+
+      assert.rejects(
+        permController.validatePermittedAccounts({}),
+        ERRORS.validatePermittedAccounts.invalidParam(),
+        'should throw on non-array object'
+      )
+    })
+
+    it('throws error on empty array of accounts', async function () {
+
+      assert.rejects(
+        permController.validatePermittedAccounts([]),
+        ERRORS.validatePermittedAccounts.invalidParam(),
+        'should throw on empty array'
+      )
+    })
+
+    it('throws error if any account value is not in keyring', async function () {
+
+      assert.rejects(
+        permController.validatePermittedAccounts([DUMMY_ACCOUNT]),
+        ERRORS.validatePermittedAccounts.nonKeyringAccount(DUMMY_ACCOUNT),
+        'should throw on non-keyring account'
+      )
+
+      assert.rejects(
+        permController.validatePermittedAccounts(ACCOUNT_ARRAYS.a.concat(DUMMY_ACCOUNT)),
+        ERRORS.validatePermittedAccounts.nonKeyringAccount(DUMMY_ACCOUNT),
+        'should throw on non-keyring account with other accounts'
+      )
+    })
+
+    it('succeeds if all accounts are in keyring', async function () {
+
+      assert.doesNotReject(
+        permController.validatePermittedAccounts(ACCOUNT_ARRAYS.a),
```

Comment:
It's a bit confusing that you're using `ACCOUNT_ARRAYS.a` here for all accounts, where as `a` and `b` are the first and second account respectively - particularly since the keys `a`, `b`, and `c` were used earlier for distinct cases/domains for testing other functions. It seems entirely incidental that `a` has all accounts.

Could you just reference the keyring accounts array directly? That would be much easier to understand - this `ACCOUNTS_ARRAY` structure doesn't seem helpful here.

Reply:
Done in 565833a